## Forecasting Vaccine Uptake: A Data-Driven Approach

![image](H1N1.jpg)

## Author : MIKE KIPTOCH

## NoteBook Structure
i) Business Understanding<br>
ii) Data Understanding<br>
iii) Data Preparation<br>
iv) Modeling<br>
v) Evaluation<br>
vi) Conclusion<br>
vii) Recommendation<br>
v) Next Steps<br>

## Business Understanding

Vaccination stands as one of humanity's greatest achievements in public health, having effectively eradicated or controlled numerous diseases. However, recent years have witnessed a concerning rise in vaccine skepticism, leading to declining immunization rates and outbreaks of preventable diseases. Understanding the factors driving vaccine hesitancy and predicting vaccination uptake is imperative for public health officials and policymakers.

This project leverages data from the National Flu Survey (NHFS 2009) to predict individuals' likelihood of receiving the H1N1 flu vaccine. By examining past vaccination patterns, the study aims to shed light on contemporary vaccination behaviors, particularly pertinent in the context of emerging pandemics like COVID-19.

## Problem Statement

Vaccine hesitancy poses a significant challenge to public health efforts, leading to decreased immunization rates and increased vulnerability to infectious diseases. Understanding the factors influencing individuals' decisions regarding vaccine uptake is crucial for designing effective interventions and promoting community immunity. In this context, the project aims to predict the likelihood of individuals receiving the H1N1 flu vaccine using machine learning techniques and data from the National Flu Survey (NHFS 2009).

## Objectives

- Prediction: Develop machine learning models to predict individuals' H1N1 vaccine uptake based on demographic, socio-economic, and attitudinal factors.

- Identify Influential Factors: Determine the most influential factors affecting H1N1 vaccine acceptance, including doctor recommendations, health insurance coverage, perceptions of vaccine effectiveness, and risk perceptions related to H1N1.

- Model Evaluation: Assess the performance of different machine learning algorithms, including Decision Tree Classifier, Logistic Regression, Random Forest, K-Nearest Neighborhood Classifier, Gradient Boosting Classifier, and XG Boosting Classifier, in predicting H1N1 vaccine uptake.

- Impact Analysis: Analyze the implications of the predictive models and identify actionable insights for public health professionals and policymakers to improve vaccination rates.

## Description of Data Source

The dataset utilized in this project comprises 10 columns and 26,708 rows, sourced from Kaggle. This dataset originates from the National Flu Survey (NHFS) of 2009, a comprehensive survey conducted to understand public attitudes and behaviors regarding influenza vaccination, specifically targeting the H1N1 flu vaccine.

## Data Preparation

In [32]:
# Importing Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt







import warnings
warnings.filterwarnings("ignore")
%matplotlib inline

In [33]:
# Load the dataset

df= pd.read_csv('H1N1_Flu_Vaccines.csv')

In [34]:
#display the 10  rows
df.head(10)

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   
5              5           3.0             1.0                        0.0   
6              6           0.0             0.0                        0.0   
7              7           1.0             0.0                        0.0   
8              8           0.0             2.0                        0.0   
9              9           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   
5                   1.0                   0.0                    1.0   
6                   0.0                   0.0                    0.0   
7                   1.0                   0.0                    1.0   
8                   1.0                   0.0                    1.0   
9                   1.0                   0.0                    0.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   
5                          0.0                      0.0   
6                          0.0                      0.0   
7                          0.0                      0.0   
8                          1.0                      1.0   
9                          1.0                      0.0   

   behavioral_touch_face  ...  rent_or_own   employment_status  \
0                    1.0  ...          Own  Not in Labor Force   
1                    1.0  ...         Rent            Employed   
2                    0.0  ...          Own            Employed   
3                    0.0  ...         Rent  Not in Labor Force   
4                    1.0  ...          Own            Employed   
5                    1.0  ...          Own            Employed   
6                    0.0  ...          Own            Employed   
7                    1.0  ...          Own            Employed   
8                    1.0  ...          Own            Employed   
9                    1.0  ...          Own  Not in Labor Force   

   hhs_geo_region                census_msa  household_adults  \
0        oxchjgsf                   Non-MSA               0.0   
1        bhuqouqj  MSA, Not Principle  City               0.0   
2        qufhixun  MSA, Not Principle  City               2.0   
3        lrircsnp       MSA, Principle City               0.0   
4        qufhixun  MSA, Not Principle  City               1.0   
5        atmpeygn       MSA, Principle City               2.0   
6        qufhixun  MSA, Not Principle  City               0.0   
7        bhuqouqj                   Non-MSA               2.0   
8        bhuqouqj  MSA, Not Principle  City               1.0   
9        qufhixun  MSA, Not Principle  City               0.0   

   household_children  employment_industry  employment_occupation  \
0                 0.0                  NaN                    NaN   
1                 0.0             pxcmvdjn               xgwztkwe   
2                 0.0             rucpziij               xtkaf

In [35]:
# check the shape
df.shape

(26707, 38)

In [36]:
# check data info

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26707 non-null  int64  
 1   h1n1_concern                 26615 non-null  float64
 2   h1n1_knowledge               26591 non-null  float64
 3   behavioral_antiviral_meds    26636 non-null  float64
 4   behavioral_avoidance         26499 non-null  float64
 5   behavioral_face_mask         26688 non-null  float64
 6   behavioral_wash_hands        26665 non-null  float64
 7   behavioral_large_gatherings  26620 non-null  float64
 8   behavioral_outside_home      26625 non-null  float64
 9   behavioral_touch_face        26579 non-null  float64
 10  doctor_recc_h1n1             24547 non-null  float64
 11  doctor_recc_seasonal         24547 non-null  float64
 12  chronic_med_condition        25736 non-null  float64
 13  child_under_6_mo

In [37]:
#checking value counts
vaccine_counts_h1n1 = df['h1n1_vaccine'].value_counts()
vaccine_counts_seasonal = df['seasonal_vaccine'].value_counts()

print("Value Counts for h1n1_vaccine:\n", vaccine_counts_h1n1)
print("\nValue Counts for seasonal_vaccine:\n", vaccine_counts_seasonal)


Value Counts for h1n1_vaccine:
 h1n1_vaccine
0    21033
1     5674
Name: count, dtype: int64

Value Counts for seasonal_vaccine:
 seasonal_vaccine
0    14272
1    12435
Name: count, dtype: int64


H1N1 Flu Vaccine:

- 21,033 respondents did not receive the H1N1 flu vaccine.
- 5,674 respondents received the H1N1 flu vaccine.
Seasonal Flu Vaccine:

- 14,272 respondents did not receive the seasonal flu vaccine.
- 12,435 respondents received the seasonal flu vaccine.

- The value counts indicate that the distribution of respondents across the "seasonal_vaccine" classes is more balanced compared to the distribution across the "h1n1_vaccine" classes.

In [38]:
# description of data
df.describe()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
count   26707.000000  26615.000000    26591.000000               26636.000000   
mean    13353.000000      1.618486        1.262532                   0.048844   
std      7709.791156      0.910311        0.618149                   0.215545   
min         0.000000      0.000000        0.000000                   0.000000   
25%      6676.500000      1.000000        1.000000                   0.000000   
50%     13353.000000      2.000000        1.000000                   0.000000   
75%     20029.500000      2.000000        2.000000                   0.000000   
max     26706.000000      3.000000        2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  ...  opinion_h1n1_vacc_effective  \
count           26579.000000  ...                 26316.000000   
mean                0.677264  ...                     3.850623   
std                 0.467531  ...                     1.007436   
min                 0.000000  ...                     1.000000   
25%                 0.000000  ...                     3.000000   
50%                 1.000000  ...                     4.000000   
75%                 1.000000  ...                     5.000000   
max                 1.000000  ...                     5.000000   

       opinion_h1n1_risk  opinion_h1n1_sick_from_vacc  \
count       26319.000000                 26312.000000   
mean            2.342566                     2.357670   
std             1.285539                     1.362766   
min             1.000000                     1.000000   
25%             1.000000                     1.000000   
50%             2.000000                     2.000000   
75%             4.000000                     4.000000   
max             5.000000                     5.000000   

       opinion_seas_vacc_effective  opinion_seas_risk  \
count                 26245.000000       26193.000000   
mean                      4.025986           2.719162   
std                       1.086565           1.385055   
min                       1.000000           1.000000   
25%                       4.000000           2.000000   
50%                       4.000000           2.000000   
75%                       5.000000           4.000000   
max                       5.000000           5.000000   

       opinion_seas_sick_from_vacc  household_adults  household_children  \
count                 26170.000000      26458.000000        26458.000000   
mean                      2.118112          0.886499            0.534583   
std                       1.332950          0.753422            0.928173   
min                       1.000000          0.000000            0.000000   
25%                  

## Data Cleaning

#### Check Null Values

In [39]:
columns_with_missing = df.isnull().sum()
columns_with_missing

respondent_id                      0
h1n1_concern                      92
h1n1_knowledge                   116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_h1n1                2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_h1n1_vacc_effective      391
opinion_h1n1_risk                388
opinion_h1n1_sick_from_vacc      395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4423
m